File: apps/server/src/server/routes/health.ts
// GET /api/health — the one public route. Also what `perch status` polls to decide whether the
// server in server.json is still alive.

import type { Hono } from "hono";
import type { ServerInfo } from "@perch/protocol";
import { configDir, defaultQueriesDir } from "../../storage/index.js";
import type { RouteDeps } from "../create-server.js";

export function registerHealthRoutes(app: Hono, deps: RouteDeps): void {
  app.get("/api/health", (c) => {
    const info: ServerInfo = {
      name: "perch",
      version: deps.version,
      pid: process.pid,
      url: deps.url ?? new URL(c.req.url).origin,
      startedAt: deps.startedAt,
      configDir: configDir(),
      queriesDir: defaultQueriesDir(),
    };
    return c.json(info);
  });
}

File: apps/server/server/health.go
package server

import (
	"net/http"
	"os"

	"perch/httpx"
)

// The one public route, and what `perch status` polls to decide whether the server in
// server.json is still alive.
func (s *Server) registerHealth() {
	s.handle("GET /api/health", func(w http.ResponseWriter, r *http.Request) error {
		info := s.ServerInfo(os.Getpid())
		if info.URL == "" {
			info.URL = requestOrigin(r)
		}
		return httpx.JSON(w, info)
	})
}

func requestOrigin(r *http.Request) string {
	scheme := "http"
	if r.TLS != nil {
		scheme = "https"
	}
	return scheme + "://" + r.Host
}
